src/slasHelper/slasHelper.ts

```diff
@@ -0,0 +1,263 @@
+  const urlSafe = (input: string) =>
+    input.replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
+
+  const crypto = await import("crypto");
+  const challenge = urlSafe(
+    crypto.default.createHash("sha256").update(codeVerifier).digest("base64")
+  );
+
+  /* istanbul ignore next */
+  if (challenge.length === 0) {
+    throw new Error("Problem generating code challenge");
+  }
+
+  return challenge;
+};
+
+/**
+ * Wrapper for the authorization endpoint. For federated login (3rd party IDP non-guest), the caller should redirect the user to the url in the url field of the returned object. The url will be the login page for the 3rd party IDP and the user will be sent to the redirectURI on success. Guest sessions return the code and usid directly with no need to redirect.
+ * @param slasClient - a configured instance of the ShopperLogin SDK client
+ * @param codeVerifier - random string created by client app to use as a secret in the request
+ * @param parameters - Request parameters used by the `authorizeCustomer` endpoint.
+ * @param parameters.redirectURI - the location the client will be returned to after successful login with 3rd party IDP. Must be registered in SLAS.
+ * @param parameters.hint? - optional string to hint at a particular IDP. Guest sessions are created by setting this to 'guest'
+ * @param parameters.usid? - optional saved SLAS user id to link the new session to a previous session
+ * @returns login url, user id and authorization code if available
+ */
+export async function authorize(
+  slasClient: ShopperLogin,
+  codeVerifier: string,
+  parameters: {
+    redirectURI: string;
+    hint?: string;
+    usid?: string;
+  }
+): Promise<{ code: string; url: string; usid: string }> {
+  const codeChallenge = await generateCodeChallenge(codeVerifier);
+
+  // Create a copy to override specific fetchOptions
+  const slasClientCopy = new ShopperLogin(slasClient.clientConfig);
+
+  // set manual redirect on server since node allows access to the location
+  // header and it skips the extra call. In the browser, only the default
+  // follow setting allows us to get the url.
+
+  // TODO: pass redirect manual as a fetch option
+
+  // slasClientCopy.clientConfig.fetchOptions = {
+  //   ...slasClient.clientConfig.fetchOptions,
+  //   redirect: 'manual',
+  // };
+
+  const options = {
+    parameters: {
+      client_id: slasClient.clientConfig.parameters.clientId,
+      code_challenge: codeChallenge,
+      ...(parameters.hint && { hint: parameters.hint }),
+      organizationId: slasClient.clientConfig.parameters.organizationId,
+      redirect_uri: parameters.redirectURI,
+      response_type: "code",
+      ...(parameters.usid && { usid: parameters.usid }),
+    },
+  };
+
+  const response = await slasClientCopy.authorizeCustomer(options, true);
+  const redirectUrl = response.headers?.get("location") || response.url;
+
+  // url is a read only property we unfortunately cannot mock out using nock
```

Comment:
But aren't we passing it as a function parameter? Can't we just pass in a falsy value?